apps/pp/views/references.py

```diff
@@ -137,19 +137,27 @@ def annotate_fetched_queryset(self, queryset):
         for reference in queryset:
             feedback = reference_to_feedback.get(reference.id)
             if feedback:
+                # TODO: setting useful and objection attrs duplicates corresponding relationships, consider removing
                 reference.useful = feedback.useful
                 reference.objection = feedback.objection
                 reference.does_belong_to_user = reference.id in user_reference_ids
+            reference.user_feedback = feedback
         return queryset
 
-    def preserialize_queryset(self, queryset):
+    def pre_serialize_queryset(self, queryset):
         data_list = []
         for reference in queryset:
-            attributes_serializer = ReferenceListSerializer.Attributes(reference, context={'request': self.request})
-            data = {'id': reference.id, 'type': 'references', 'attributes': attributes_serializer.data}
-            set_relationship(data, reference, attr='reference_request_id')
-            set_relationship(data, reference.user, attr='id')
-            data_list.append(data)
+            data = {'attributes': reference}
+            pre_serializer = DataPreSerializer(reference, data)
+            pre_serializer.set_relation(get_resource_name(reference, related_field='reference_request_id'),
+                                        resource_id=reference.reference_request_id)
+            pre_serializer.set_relation(get_resource_name(reference.user),
+                                        resource_id=reference.user.id)
+            pre_serializer.set_relation(get_resource_name(reference.user_feedback, model=UserReferenceFeedback),
+                                        resource_id=reference.user_feedback)
+            pre_serializer.set_relation(get_resource_name(reference, related_field='reference_reports'),
+                                        resource_id=reference.user_reference_reports)
+            data_list.append(pre_serializer.data)
```

Comment:
It i not immediately clear to me, what arguments are needed for `get_resource_name` in each case. Maybe we could make a short guide? Or write aliases for different use cases?
(It is late at night too, but still I think it's tough to figure out the application rule by yourself)

Reply:
There was some hundred of versions of this util before I managed to simplify it to this point, but I understand it's not that simple, so I'll take another try to go further with it.